From 790f73acb37aff29a4791d660fb898433f5c2f49 Mon Sep 17 00:00:00 2001
From: "p.tumakaev" <p.tumakaev@omprussia.ru>
Date: Mon, 29 Jun 2020 14:29:52 +0300
Subject: [PATCH] Skip invalid WatchId in geolocation.clearWatch()

ClearWatch() stores passed WatchId in mClearedWatchIDs without checking
if it was previously generated by WatchPosition(). As a result, all
subsequent WatchPosition() with such invalid identifiers are automatically
canceled at the "Allow" stage.
---
 dom/geolocation/nsGeolocation.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dom/geolocation/nsGeolocation.cpp b/dom/geolocation/nsGeolocation.cpp
index 0969cd771dc5..a25172db45db 100644
--- a/dom/geolocation/nsGeolocation.cpp
+++ b/dom/geolocation/nsGeolocation.cpp
@@ -1354,7 +1354,7 @@ Geolocation::WatchPosition(GeoPositionCallback aCallback,
 NS_IMETHODIMP
 Geolocation::ClearWatch(int32_t aWatchId)
 {
-  if (aWatchId < 0) {
+  if (aWatchId < 0 || aWatchId >= mLastWatchId) {
     return NS_OK;
   }
 
-- 
2.17.1

